mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SimpleMessageSearchIndex.java

```diff
@@ -68,6 +70,40 @@
  *
  */
 public class SimpleMessageSearchIndex implements MessageSearchIndex {
+
+    private static class SearchComparator implements Comparator<MailboxMessage> {
+        public static final int REVERSE = -1;
+        public static final int EQUALS = 0;
+        private final List<SearchQuery.Sort> sorts;
+
+        public SearchComparator(SearchQuery query) {
+            sorts = query.getSorts();
+        }
+
+        @Override
+        public int compare(MailboxMessage message, MailboxMessage other) {
+            return sorts.stream()
+                .mapToInt(sort -> {
+                    SearchQuery.Sort.SortClause sortClause = sort.getSortClause();
+                    int sortResult;
+                    switch (sortClause) {
+                        case Arrival:
+                            sortResult = message.metaData().getInternalDate().compareTo(other.getInternalDate());
+                            break;
+                        default:
+                            sortResult = message.compareTo(other);
+                            break;
+                    }
+                    if (sort.isReverse()) {
+                        return sortResult * REVERSE;
+                    }
+                    return sortResult;
+                })
+                .findFirst()

```

Comment:
you need to filter out 0s in order to proceed to the next sort in case of equality.